fix: clone volume content to requested volume

acortelyou opened the pull request:
/kind bug

**What this PR does / why we need it**:

`CreateVolume()` now follows the established volume provisioning logic when cloning volume content.

This ensures the created volume has the expected configuration, metadata and content.

**Which issue(s) this PR fixes**:

- Resolves issue #1495
- Prevents SAS tokens from being logged in clear text during volume cloning
- Avoids duplication of azcopy jobs when `containerName` or `containerNamePrefix` are specified

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  <!-- Common commit types:
        build: Build 🏭
        chore: Maintenance 🔧
        ci: Continuous Integration 💜
        docs: Documentation 📘
        feat: Features 🌈
        fix: Bug Fixes 🐞
        perf: Performance Improvements 🚀
        refactor: Code Refactoring 💎
        revert: Revert Change ◀️
        style: Code Style 🎶
        security: Security Fix 🛡️
        test: Testing 💚 -->
- [ ] includes documentation
- [x] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:

- Volumes are now provisioned correctly when cloning

andyzhangx commented:
/ok-to-test

acortelyou commented:
/retest

andyzhangx commented:
@acortelyou I have pushed a new commit(https://github.com/kubernetes-sigs/blob-csi-driver/pull/1504/commits/b40c5f9a521cac68574d3d3c3307d5907a53e7d5), only copy volume to different account if storageAccount is specified

acortelyou commented:
Is some code missing from the commit? Or did you want me to implement?

andyzhangx commented:
> Is some code missing from the commit? Or did you want me to implement?

@acortelyou I have made the changes already, thx

acortelyou commented:
I have some concerns about these changes, will review more thoroughly tomorrow to make sure I'm not missing something. In short, if I don't specify an account in the volume request storage class I would expect the destination to be dynamically provisioned the exact same way it would be if there was not a data source specified. I have no expectation for specifying a data *source* to influence the data *destination*.

andyzhangx commented:
> I have some concerns about these changes, will review more thoroughly tomorrow to make sure I'm not missing something. In short, if I don't specify an account in the volume request storage class I would expect the destination to be dynamically provisioned the exact same way it would be if there was not a data source specified. I have no expectation for specifying a data _source_ to influence the data _destination_.

@acortelyou the original behavior is copying the volume into the same account as source volume, we should keep the same behavior unless user specifies `storageAccount` in storage class. Also if we copy the volume to the account with dynamic matched by the driver, it's difficult for user to find out which destination account has the copied volume since the driver has it's own logic to search a matching account. that's the reason why we only supports copying the volume to the source account in the beginning.

acortelyou commented:
Thanks @andyzhangx, I appreciate your quick responses and willingness to collaborate on a solution.

I believe the user expects their volume to be provisioned as configured whether or not they provide a `datasource`.



> the original behavior is copying the volume into the same account as source volume

Correct, this is a bug and contradicts the documentation.

> we should keep the same behavior unless user specifies `storageAccount` in storage class

I disagree. We should have the volume provisioning behavior always match the documentation and user intent.

> it's difficult for user to find out which destination account has the copied volume

The documentation provides clear guidance on where/how volumes are to be provisioned and also provides the means to locate and co-locate the volumes if desired.

```
parameters:
  subscriptionID: >
    Specify Azure subscription ID where blob storage directory will be created.
  resourceGroup: >
    If empty, driver will use the same resource group name as current cluster.
  storageAccount: >
    When a specific storage account name is not provided, the driver will look for a suitable storage account that matches the account settings within the same resource group. 
    If it fails to find a matching storage account, it will create a new one. 
    However, if a storage account name is specified, the storage account must already exist.
  containerName: >
    If empty, driver creates a new container name, starting with pvc-fuse for blobfuse or pvc-nfs for NFS v3.
  containerNamePrefix: >
    Specify Azure storage directory prefix created by driver.
  tags: >
    Specify tags for storage account created by the driver.
  matchTags: >
    Match tags when driver tries to find a suitable storage account.
  skuName: >
    Specify skuName for storage account.
  location: >
    Specify location for storage account.
```

https://learn.microsoft.com/en-us/azure/aks/azure-csi-blob-storage-provision#storage-class-parameters-for-dynamic-persistent-volumes

I suspect we'll want to refactor some of these changes so that we can fail faster before potentially creating an account or container, but it would require additional modifications to the tests.

Thanks!

FYI I pulled the logs for the last month and it looks like no-one other than my team and the E2EAKS runners are even trying to use blob volume cloning. I suspect it's because the current behavior is so incredibly limiting. After fixing this bug, blob volume cloning will be extremely useful for a bunch of scenarios. Blobfuse2 can be way more performant than the other csi options we've benchmarked and we want to see it succeed.

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -781,9 +779,9 @@ func (d *Driver) copyBlobContainer(req *csi.CreateVolumeRequest, accountSasToken
+		klog.V(2).Infof("copy blob container %s to %s", srcPath, dstPath)
 		execFunc := func() error {
-			cmd := exec.Command("azcopy", "copy", srcPath, dstPath, "--recursive", "--check-length=false")
+			cmd := exec.Command("azcopy", "copy", srcPath+accountSasToken, dstPath+accountSasToken, "--recursive", "--check-length=false")
```
this is using the same account sas token, if source account and dest account are different, it would break.

andyzhangx replied:
should get srcAccountSASToken if it's two accounts, e.g. 
```
	srcAccountSasToken := dstAccountSasToken
	if srcAccountName != dstAccountName && dstAccountSasToken != "" {
		srcAccountOptions := &azure.AccountOptions{
			Name:                srcAccountName,
			ResourceGroup:       accountOptions.ResourceGroup,
			SubscriptionID:      accountOptions.SubscriptionID,
			GetLatestAccountKey: accountOptions.GetLatestAccountKey,
		}
andyzhangx marked this conversation as resolved.
		if srcAccountSasToken, _, err = d.getAzcopyAuth(ctx, srcAccountName, "", storageEndpointSuffix, srcAccountOptions, nil, "", secretNamespace, true); err != nil {
			return err
		}
	}
```

acortelyou replied:
I pushed changes which better encapsulate the volume content source validation and copy, including src sas token and associated tests. Please check that `secrets`, `secretName`, `secretNamespace` logic is correct.

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -435,19 +442,19 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 	}
 
+	klog.V(2).Infof("begin to create container(%s) on account(%s) type(%s) subsID(%s) rg(%s) location(%s) size(%d)", validContainerName, accountName, storageAccountType, subsID, resourceGroup, location, requestGiB)
+	if err := d.CreateBlobContainer(ctx, subsID, resourceGroup, accountName, validContainerName, secrets); err != nil {
```
azcopy would create container before volume clone, so let's revert this

acortelyou replied:
My intent here was for `CreateBlobContainer()` to configure `Access`, `PublicAccess`, `Metadata` uniformly. I see metadata on azcopy but not public access.

acortelyou replied:
ah, i see that actually an account setting not container setting i will make the change with metadata includes

acortelyou replied:
I have slight reservations still about not creating the blob container first with `CreateBlobContainer()` because azcopy can't default the blob container to private, so there's a slight difference in behavior.

However, since the default for the storage container is private this should only change the outcome when using a static storage account with public access turned on or when explicitly requested.

I left it as a separate commit so that you can decide whether that's a concern and either revert or squash.

acortelyou replied:
Ah, nope, the sanity check test threw this `failed to parse user input due to error: content-type, content-encoding, content-language, content-disposition, cache-control, or metadata is not supported while copying from service to service\n",`. I'll revert the change.

acortelyou commented:
/retest

umagnus commented on pkg/blob/controllerserver.go:
```diff
@@ -367,9 +367,15 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 
+	secrets := req.GetSecrets()
+
+	volumeContentSourceSASToken, volumeContentSourcePath, err := d.validateVolumeContentSource(ctx, req, storageEndpointSuffix, secrets, secretName, secretNamespace, getLatestAccountKey)
```
I think we don't need to add a function `validateVolumeContentSource` here. We can  move it to `copyBlobContainer` and if `srcAccountName != dstAccountName`, then generate `srcSASToken`, otherwise we just need to generate sas token once.

acortelyou replied:
The intent was to confirm we have have access to the volume source before provisioning a volume we can't populate. I could add a check to skip the sas generation for the *destination* if it matches the source. The added function was just to isolate the changes and make it easier to test, I can fold it back in if you feel strongly about it.

acortelyou replied:
@umagnus on review, the generated sas token is cached. there is no need for an additional optimization.

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -367,9 +367,15 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 
+	secrets := req.GetSecrets()
+
+	volumeContentSourceSASToken, volumeContentSourcePath, err := d.validateVolumeContentSource(ctx, req, storageEndpointSuffix, secrets, secretName, secretNamespace, getLatestAccountKey)
```
pls only call `validateVolumeContentSource` in the `if req.GetVolumeContentSource() != nil { ..}` logic, that would more clear logic

acortelyou commented:
@umagnus I have removed the extraneous `funcs` and added a check before regenerating the azcopy auth env.
@andyzhangx I have added `if volContentSource != nil` around all appropriate code blocks and comments to show the logical code flow.

One now-redundant test has been removed.
I have reverted most non-essential changes and scoped this PR tightly to the issue at hand.

acortelyou commented:
I think I spotted a volume cloning bug where `GetAzcopyJob` was grep'ing using the request `volName` instead of the actual destination `validContainerName`, which could result in multiple duplicate azcopy jobs being started.

I left that fix as a separate commit so that it can be easily reverted if the existing behavior was intentional.

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -67,6 +67,9 @@ const (
 // CreateVolume provisions a volume
 func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+
+	// Request
```
pls remove these comments and new empty lines

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -81,19 +84,14 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
+	volContentSource := req.GetVolumeContentSource()
+	secrets := req.GetSecrets()
+
+	// Parameters
```
pls remove these comments

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -367,9 +378,53 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 		mc.ObserveOperationWithResult(isOperationSucceeded, VolumeID, volumeID)
 	}()
 
+	// Volume Source
```
remove

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -367,9 +378,53 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
+		}
+	}
+
+	// Storage Account
```
remove

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -423,33 +478,32 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
-		}
-		validContainerName = getValidContainerName(validContainerName, protocol)
-		setKeyValueInMap(parameters, containerNameField, validContainerName)
+	// Blob Container
```
remove

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -423,33 +478,32 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
-		accountSASToken, authAzcopyEnv, err := d.getAzcopyAuth(ctx, accountName, accountKey, storageEndpointSuffix, accountOptions, secrets, secretName, secretNamespace)
-		if err != nil {
-			return nil, status.Errorf(codes.Internal, "failed to getAzcopyAuth on account(%s) rg(%s), error: %v", accountOptions.Name, accountOptions.ResourceGroup, err)
+	// Volume Content
```
remove

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -466,6 +520,8 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
 		}
 	}
 
+	// Response
```
remove

umagnus commented on pkg/blob/controllerserver.go:
```diff
@@ -367,9 +378,53 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
+		}
+		return nil, status.Errorf(codes.Aborted, volumeOperationAlreadyExistsFmt, volName)
+	}
+	defer d.volumeLocks.Release(volName)
```
This part can be moved to L362

andyzhangx commented on pkg/blob/controllerserver.go:
```diff
@@ -423,31 +467,24 @@ func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest)
+		dstAzcopyAuthEnv := srcAzcopyAuthEnv
+		dstAccountSASToken := srcAccountSASToken
+		if srcSubscriptionID != subsID || srcResourceGroupName != resourceGroup || srcAccountName != accountName {
+			if dstAccountSASToken, dstAzcopyAuthEnv, err = d.getAzcopyAuth(ctx, accountName, accountKey, storageEndpointSuffix, accountOptions, secrets, secretName, secretNamespace); err != nil {
```
`if volContentSource != nil`, I think `dstAccountSASToken, dstAzcopyAuthEnv, err = d.getAzcopyAuth` could be done before createBlobContainer, that would be better

andyzhangx replied:
make all the auth check before createBlobContainer

acortelyou commented:
/retest

acortelyou commented:
/retest

acortelyou commented:
/retest

andyzhangx commented:
/retest

andyzhangx reviewed (approved):
/lgtm
